ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule.hs

```diff
@@ -558,13 +607,171 @@ splitFragmentAtSlotNo slotNo (fragment :> block) =
 splitFragmentAtSlotNo _ (Empty anchor) =
   (Empty anchor, Empty anchor)
 
+data RollbackMode =
+  RollbackOneBranch
+  |
+  RollbackRepeatBranch
+  deriving (Eq, Show)
+
+-- | Generate a peer schedule that rolls back repeatedly.
+--
+-- See 'rollbackSpamPointSchedule'.
+rollbackSpamPeerSingle ::
+  Bool ->
+  (TestFrag, (TestFrag, (TestFrag, TestFrag))) ->
+  [NodeState]
+rollbackSpamPeerSingle bulk (_, (prefix, (fork, suffix))) =
+  (prefixStates ++ join (zipWith (++) forkStates suffixStates))
+  where
+    prefixStates = banalStates prefix
+    suffixStates = rollback (banalStates suffix)
+    forkStates = rollback (banalStates fork)
+    rollback frag | bulk = pure <$> frag
+                  | otherwise = tails frag
+
+rollbackSpamPointScheduleSingle ::
+  Peers Int ->
+  Bool ->
+  BlockTree TestBlock ->
+  Maybe (PointSchedule, Maybe a)
+rollbackSpamPointScheduleSingle freqs bulk blockTree =
+  fmap (\ a -> (a, Nothing)) (interleaveWithFrequencies freqs peers)
+  where
+    peers = bimapPeers (banalStates . fst) (rollbackSpamPeerSingle bulk) (blockTreePeersFull blockTree)
+
+-- | First serve the prefix, advertising the honest tip.
+-- Then serve k blocks from the fork, advertising the fork's tip.
+-- Then move the fork's anchor one block ahead, rewriting all the hashes and slot numbers.
+-- Serve the extra honest block, then serve k blocks off the shifted fork, advertising the new fork tip for both.
+-- Repeat until the honest tip is the anchor.
+rollbackSpamPeerRepeat ::
```

Comment:
Reusing alternative chains like this without breaking expectations about competing chains is going to be some effort. I would think we don't need to modify the generated chains like this.

Do you still prefer to have this implementation merged?

Reply:
no, I don't think that would be sensible to keep